File: leetcode/python/word_search_I_II_bt.py
__author__ = 'BeyondSky'

from trie import Trie
from trie import TrieNode

class WordsearchI(object):

    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        row = len(board)
        col = len(board[0])
        if row == 0 or col == 0:
            return False
        if row * col < len(word):
            return False

        visited = [ [0]*col for y in range(row) ]
        for i in range(row):
            for j in range(col):
                ret = self.word_search_helper(board, word, visited, i, j, 0)
                if ret is True:
                    return True;

        return False;


    def word_search_helper(self, board, w, visited, i, j, pos):
        if board[i][j] != w[pos]:
            return False

        if pos == len(w) - 1:
            return True

        row = len(board)
        col = len(board[0])

        pos += 1
        visited[i][j] = 1

        if i > 0 and visited[i-1][j] == 0 and self.word_search_helper(board, w, visited, i-1, j, pos):
            return True
        if i < row-1 and visited[i+1][j] == 0 and self.word_search_helper(board, w, visited, i+1, j, pos):
            return True
        if j > 0 and visited[i][j-1] == 0 and self.word_search_helper(board, w, visited, i, j-1, pos):
            return True
        if j < col-1 and visited[i][j+1] == 0 and self.word_search_helper(board, w, visited, i, j+1, pos):
            return True


        visited[i][j] = 0
        return False

class WordsearchII?:
    hs = set()

    def findWords(self, board, words):
        """
        :type board: List[List[str]]
        :type words: List[str]
        :rtype: List[str]
        """
        res = []
        row = len(board)
        col = len(board[0])
        if row == 0 and col == 0:
            return res

        visited = [[0] * col for x in range(row)]

        tr = Trie()
        for s in words:
            tr.add_word(s)

        root = tr.root

        for i in range(row):
            for j in range(col):
                # prune search earlier if word is not in trie
                idx = board[i][j]
                if root.children.get(idx) is None:
                    continue

                visited[i][j] = 1
                self.word_search_helper2_dfs(board, res, root.children[idx], visited, i, j)
                visited[i][j] = 0

        return sorted(res)

    def word_search_helper2_dfs(self, board, res, trie_node, visited, i, j):
        if trie_node.isWord is True:
            s = trie_node.get_word()
            if s not in self.hs:    # avoid dup to the res list
                self.hs.add(s)
                res.append(s)

        row = len(board)
        col = len(board[0])

        # UP
        if i > 0 and visited[i-1][j] == 0:
            idx = board[i-1][j]
            if trie_node.children.get(idx) is not None:
                visited[i-1][j] = 1
                self.word_search_helper2_dfs(board, res, trie_node.children[idx], visited, i-1, j)
                visited[i-1][j] = 0

        # DOWN
        if i < row-1 and visited[i+1][j] == 0:
            idx = board[i+1][j]
            if trie_node.children.get(idx) is not None:
                visited[i+1][j] = 1
                self.word_search_helper2_dfs(board, res, trie_node.children[idx], visited, i+1, j)
                visited[i+1][j] = 0

        # LEFT
        if j > 0 and visited[i][j-1] == 0:
            idx = board[i][j-1]
            if trie_node.children.get(idx) is not None:
                visited[i][j-1] = 1
                self.word_search_helper2_dfs(board, res, trie_node.children[idx], visited, i, j-1)
                visited[i][j-1] = 0

        # RIGHT
        if j < col-1 and visited[i][j+1] == 0:
            idx = board[i][j+1]
            if trie_node.children.get(idx) is not None:
                visited[i][j+1] = 1
                self.word_search_helper2_dfs(board, res, trie_node.children[idx], visited, i, j+1)
                visited[i][j+1] = 0


if __name__ == "__main__":
    outer = WordsearchII()
    matrix1 = [ ['a', 'b', 'd'], ['e', 'c', 'f']]
    matrix2 = ["a", "a"]
    print(outer.findWords(matrix2, ['a']))

File: leetcode/java/AlgoBackTracking/WordSearchI_II_BT.java
package AlgoBackTracking;

import DataStructureTreeTrie.TrieNodeArray;
import DataStructureTreeTrie.TrieTreeArray;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;


public class WordSearchI_II_BT {
    /*
	 * Given a 2D board and a list of words from the dictionary, find all words in the board.
	 * Each word must be constructed from letters of sequentially adjacent cell, where "adjacent" cells are those
	 * horizontally or vertically neighboring. The same letter cell may not be used more than once in a word.
		For example,
		Given words = ["oath","pea","eat","rain"] and board =

		[
		  ['o','a','a','n'],
		  ['e','t','a','e'],
		  ['i','h','k','r'],
		  ['i','f','l','v']
		]
		Return ["eat","oath"].
	 */
	
	/* HINT
	 * You would need to optimize your backtracking to pass the larger test. Could you stop backtracking earlier?
	 * If the current candidate does not exist in all words' prefix, you could stop backtracking immediately. 
	 * What kind of data structure could answer such query efficiently? Does a hash table work? 
	 * Why or why not? How about a Graph.TrieTree? If you would like to learn how to implement a basic trie,
	 * please work on this problem: Implement Graph.TrieTree (Prefix Tree) first.
	 */

    // Algorithm: save target words to TrieTree. DFS
    //  stop of current iterator if it doesn't match any prefix of TrieTree
    //  once the target word is found, remove it from TrieTree

    Set<String> hs = new HashSet<>();    // ensure result has unique string

    public List<String> findWordsII_trie(char[][] board, String[] words) {
        List<String> res = new ArrayList<>();
        int row = board.length;
        int col = board[0].length;
        if (row == 0 && col == 0) return res;
        boolean[][] visited = new boolean[row][col];

        // build TrieTree first
        TrieTreeArray tr = new TrieTreeArray();
        for (String s : words) {
            tr.addWord(s);
        }

        TrieNodeArray root = tr.root;

        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {

                // check trie to prune search earlier
                int idx = board[i][j] - 'a';
                if (root.child[idx] == null)
                    continue;

                visited[i][j] = true;
                wordsearchTrieDFS(board, res, root.child[idx], visited, i, j);    // try each any cell
                visited[i][j] = false;
            }
        }

        return res;
    }

    public void wordsearchTrieDFS(char[][] board, List<String> res, TrieNodeArray node, boolean[][] visited, int i, int j) {
        if (node.isWord) {
            String s = node.getWord();
            if (!hs.contains(s)) {
                hs.add(s);
                res.add(s);
            }
        }

        int idx;

        int row = board.length;
        int col = board[0].length;

        // UP
        if (i > 0 && !visited[i - 1][j]) {
            idx = board[i - 1][j] - 'a';
            if (node.child[idx] != null) {
                visited[i - 1][j] = true;
                wordsearchTrieDFS(board, res, node.child[idx], visited, i - 1, j);
                visited[i - 1][j] = false;
            }
        }

        // DOWN
        if (i < row - 1 && !visited[i + 1][j]) {
            idx = board[i + 1][j] - 'a';
            if (node.child[idx] != null) {
                visited[i + 1][j] = true;
                wordsearchTrieDFS(board, res, node.child[idx], visited, i + 1, j);
                visited[i + 1][j] = false;
            }
        }

        // LEFT
        if (j > 0 && !visited[i][j - 1]) {
            idx = board[i][j - 1] - 'a';
            if (node.child[idx] != null) {
                visited[i][j - 1] = true;
                wordsearchTrieDFS(board, res, node.child[idx], visited, i, j - 1);
                visited[i][j - 1] = false;
            }
        }

        // RIGHT
        if (j < col - 1 && !visited[i][j + 1]) {
            idx = board[i][j + 1] - 'a';
            if (node.child[idx] != null) {
                visited[i][j + 1] = true;
                wordsearchTrieDFS(board, res, node.child[idx], visited, i, j + 1);
                visited[i][j + 1] = false;
            }
        }
    }


    // brutal force, TLE
    public List<String> findWordSearchII_simpleTLE(char[][] board, String[] words) {
        List<String> res = new ArrayList<>();
        int row = board.length;
        int col = board[0].length;
        if (row == 0 && col == 0) return res;
        boolean[][] visited = new boolean[row][col];

        for (String s : words) {    // search word use WordSearchI()

            for (int i = 0; i < row; i++) {
                for (int j = 0; j < col; j++) {
                    visited[i][j] = false;
                }
            }

            for (int i = 0; i < row; i++) {
                for (int j = 0; j < col; j++) {
                    if (wordsearchIhelper(board, s, visited, i, j, 0)) {    // try each any cell
                        res.add(s);
                    }
                }
            }
        }
        return res;
    }

	
	/* 
	 *  Given a 2D board and a word, find if the word exists in the grid.
	 *	The word can be constructed from letters of sequentially adjacent cell, 
	 *	where "adjacent" cells are those horizontally or vertically neighboring. 
	 *	The same letter cell may not be used more than once.
		For example,	Given board =
		[
		  ["ABCE"],
		  ["SFCS"],
		  ["ADEE"]
		]
	
		word = "ABCCED", -> returns true,
		word = "SEE", -> returns true,
		word = "ABCB", -> returns false.
	 */

    // Algorithm: BT, extensive searching for neighbor element

    public boolean existWordSearchI_BT_DFS(char[][] board, String word) {
        int row = board.length;
        int col = board[0].length;
        if (row == 0 && col == 0) return false;
        if (row * col < word.length()) return false;

        boolean[][] visited = new boolean[row][col];    // each cell can only be used ONCE
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                visited[i][j] = false;
            }
        }

        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (wordsearchIhelper(board, word, visited, i, j, 0)) {    // try each any cell
                    return true;
                }
            }
        }
        return false;
    }

    public boolean wordsearchIhelper(char[][] board, String target_word, boolean[][] visited,
                                     int i, int j, int str_pos) {

        if (board[i][j] != target_word.charAt(str_pos))
            return false;

        if (str_pos == target_word.length() - 1)    // no error so far, done
            return true;

        // this char is matched on [i, j]. continue matching next using char
        str_pos++;

        // forward tracking
        visited[i][j] = true;

        int row = board.length;
        int col = board[0].length;

        // search i-1, UP
        if (i > 0 && !visited[i - 1][j] &&
                wordsearchIhelper(board, target_word, visited, i - 1, j, str_pos)) return true;
        // search i+1, DOWN
        if (i < row - 1 && !visited[i + 1][j] &&
                wordsearchIhelper(board, target_word, visited, i + 1, j, str_pos)) return true;
        // search j-1, LEFT
        if (j > 0 && !visited[i][j - 1] &&
                wordsearchIhelper(board, target_word, visited, i, j - 1, str_pos)) return true;
        // search j+1, RIGHT
        if (j < col - 1 && !visited[i][j + 1] &&
                wordsearchIhelper(board, target_word, visited, i, j + 1, str_pos)) return true;

        // all the above further search failed, backtracking
        visited[i][j] = false;
        return false;
    }
}
